pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/NumericalFilterOptimizer.java

```diff
@@ -101,6 +111,15 @@ Expression optimizeChild(Expression filterExpression, @Nullable Schema schema) {
             }
           }
         }
+
+        if (kind == FilterKind.BETWEEN) {
```

Comment:
Not introduced in this PR, but we have multiple levels of `switch` and `if` on `kind`, which is quite hard to read. We didn't check whether lower/upper bound is numeric literal here, which is different with other filter kinds. Is this intentional?

Actually we probably shouldn't check that and also allow rewrite on string range. E.g. `intCol = '1.0'` should be valid

Reply:
> We didn't check whether lower/upper bound is numeric literal here, which is different with other filter kinds. Is this intentional?

The check is in the `rewriteBetweenExpression` method itself, since we want to do the rewrite even if one of lower bound / upper bound is a numeric literal ideally (and I didn't want to add this into the common logic for other filter kinds).

> Actually we probably shouldn't check that and also allow rewrite on string range. E.g. intCol = '1.0' should be valid

Postgres doesn't allow this btw, do we still want to add support for this?

> Not introduced in this PR, but we have multiple levels of switch and if on kind, which is quite hard to read

Agreed, I've pushed a commit trying to clean this up - let me know if this looks good or if you had something else in mind.